### PR TITLE
feat(medusa-react,medusa,types,dashboard): added empty state + table for promotions list page

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds empty state for promotions list page
- lists all promotions with pagination

Screenshot 2024-03-26 at 14 19 27

Screenshot 2024-03-27 at 20 46 17

### CODE REVIEW

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import { adminPromotionKeys } from "../../../lib/api-v2"
+import { medusa, queryClient } from "../../../lib/medusa"
+
+const promotionsListQuery = () => ({
+  queryKey: adminPromotionKeys.list({ limit: 20, offset: 0 }),
+  // Question: whats the v2 equivalenet of this?
+  queryFn: async () => medusa.admin.discounts.list({ limit: 20, offset: 0 }),
+})
+
+// Question: Are we using this anywhere magically? Can this file be removed?
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts]: q: is this file needed? What exactly does it do?

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/promotions-list.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// TODO: This needs to be replaced with promotions
+// Question: Where is this being initialized?
+import after from "medusa-admin:widgets/discount/list/after"
+import before from "medusa-admin:widgets/discount/list/before"
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/promotions-list.tsx]: q: how can i create injections for promotions?

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts (hunk 1 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts]: cc. @kasperkristensen @fPolic i have added some comments in the PR if you guys can help here? Thanks ❤️

Comment (fPolic) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts]: This is a react router pattern for preloading data for the route. Each route can define a main component and a loader, data loaded here will be available in the `Component` exported from the index.

https://reactrouter.com/en/main/route/loader

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import { useDataTable } from "../../../../../hooks/use-data-table"
+import { promotionsLoader } from "../../loader"
+
+const PAGE_SIZE = 20
+
+export const PromotionListTable = () => {
+  const { t } = useTranslation()
+
+  // Question: what is this?
+  const initialData = useLoaderData() as Awaited<
```
Comment (fPolic) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx:23]: this will be the data that the loader form `loader.ts` returns

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { AdminPromotionsListRes } from "@medusajs/medusa"
+import { Response } from "@medusajs/medusa-js"
+import { QueryClient } from "@tanstack/react-query"
+
+import { adminPromotionKeys } from "../../../lib/api-v2"
+import { medusa, queryClient } from "../../../lib/medusa"
+
+const promotionsListQuery = () => ({
+  queryKey: adminPromotionKeys.list({ limit: 20, offset: 0 }),
+  // Question: whats the v2 equivalenet of this?
```
Comment (fPolic) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/loader.ts]: not sure if there is any agreed pattern atm but Oli started defining hooks under `lib/api-v2/*` [e.g. here](https://github.com/medusajs/medusa/pull/6821/files#diff-026e3394090d3730212762d5f3d113bacae5dba4c27a2df676f00a7539a9b2c0R53)

File: packages/admin-next/dashboard/src/hooks/table/filters-v2/use-promotion-table-filters.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { useTranslation } from "react-i18next"
+import { Filter } from "../../../components/table/data-table"
+
+export const usePromotionTableFilters = () => {
+  const { t } = useTranslation()
+
+  let filters: Filter[] = [
+    { label: t("fields.createdAt"), key: "created_at" },
```
Comment (sradevski) [inline packages/admin-next/dashboard/src/hooks/table/filters-v2/use-promotion-table-filters.tsx]: I think we can remove the `map` and just add `type: date` inline here :)

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    ReturnType<ReturnType<typeof promotionsLoader>>
+  >
+
+  const { searchParams, raw } = usePromotionTableQuery({ pageSize: PAGE_SIZE })
+  const { promotions, count, isLoading, isError, error } = useV2Promotions(
+    { ...searchParams },
+    {
+      initialData,
+      // Question: what is this?
+      // TODO: reenable this once its figured out
```
Comment (sradevski) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx]: I believe this tells `react-query` to not delete old data when you fetch a new batch. For example, if you are doing infinite scrolling, the second request will contain additional 20 items, but we still want to keep the old 20 in memory.

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      }),
+      confirmText: t("actions.delete"),
+      cancelText: t("actions.cancel"),
+    })
+
+    if (!res) {
+      return
+    }
+
+    await mutateAsync()
```
Comment (sradevski) [inline packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx:101]: We should handle failures here, either by setting `onError` on the useMutate hook, or using a try/catch

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx (hunk 3 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx:23]: interesting, thanks @fPolic!

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx (hunk 7 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx:101]: definitely, I've added some todos in the code, will get to this when i'm tackling delete.

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx (hunk 6 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/components/promotion-list-table/promotion-list-table.tsx]: ah, that makes sense. Thanks!

File: packages/admin-next/dashboard/src/hooks/table/query-v2/use-promotion-table-query.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+type UsePromotionTableQueryProps = {
+  prefix?: string
+  pageSize?: number
+}
+
+export const usePromotionTableQuery = ({
+  prefix,
+  pageSize = 20,
+}: UsePromotionTableQueryProps) => {
+  const queryObject = useQueryParams(
```
Comment (fPolic) [inline packages/admin-next/dashboard/src/hooks/table/query-v2/use-promotion-table-query.tsx:13]: todo: filtering and search is broken due to server errors

**Review by fPolic — COMMENTED** [review]
Looks good overall! 
TODO is to make filtering and search work.

File: packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/promotions-list.tsx (hunk 2 above)
Comment (fPolic) [reply packages/admin-next/dashboard/src/v2-routes/promotions/promotion-list/promotions-list.tsx]: think that it would be fine to postpone this for a follow up ticket since zones for v1 also need to be revisited

File: packages/admin-next/dashboard/src/hooks/table/query-v2/use-promotion-table-query.tsx (hunk 8 above)
Comment (riqwan) [reply packages/admin-next/dashboard/src/hooks/table/query-v2/use-promotion-table-query.tsx:13]: Thanks @fPolic, I have a separate todo on that one. Will be tackling this in a couple of PRs down

**Review by fPolic — APPROVED** [review]
